# Power Method

***

The eigenvalues of a n×n matrix $\mathbf{[A]}$ are obtained by solving its characteristic equation.

$\lambda^{n}$ + $a_{n-1}$$\lambda^{n-1}$ + $a_{n-2}$$\lambda^{n-2}$ + ... + $a_{0}$ = **0**

- For large values of n, polynomial equations like this one are difficult and time-consuming to solve. 
- Moreover, numerical techniques for approximating roots of polynomial equations of high degree are sensitive to rounding errors. 
- Therefore, we look at an alternative method for approximating eigenvalues.


***

## Power Method

- Like the Jacobi and Gauss-Seidel methods, the power method for approximating eigenvalues is iterative. 
- This method can be used only to find the eigenvalue of $\mathbf{[A]}$ that is largest in absolute value that is, the dominant eigenvalue of $\mathbf{[A]}$. 
- First, we assume that the matrix $\mathbf{[A]}$ has a dominant eigenvalue with corresponding dominant eigenvectors. Then we choose an initial approximation of one of the dominant eigenvectors of $\mathbf{[A]}$. 
- This initial approximation must be a nonzero vector in $R^{n}$.

## Inverse Power Method

- This method is used for calculating the smallest eigenvalue of the given matrix $\mathbf{[A]}$. 
- For a n × n matrix $\mathbf{[A]}$, with unique eigenvalue of minimum absolute magnitude can also be obtained by power method. 
- If α is dominant eigenvalue of $\mathbf{[A]}$, then $\frac{1}{|\alpha|}$ is dominant eigenvalue of $[A]^{-1}$ which is, consequently, the smallest eigenvalue of $\mathbf{[A]}$.

In [1]:
#importing libraries

import numpy as np
import sys
from pprint import pprint
import matplotlib.pyplot as plt

In [2]:
with open(r'Matrix1.txt') as f:
    a = [[int(num) for num in line.split()] for line in f]
    
a = np.array(a, float)
pprint(a)

n = len(a)

array([[-30.,  10.,  20.],
       [ 10.,  40., -50.],
       [ 20., -50., -10.]])


In [3]:
# this can be used if text file is not available

# a = np.array([[-30, 10, 20][10, 40, -50], [20, -50, -10]], float)

# n = len(a)

In [4]:
# initialising guess vector 

x = np.array([1,1,1], float)

errmax = 0.0000001

itermax = 500

In [5]:
# Power Method code for finding dominant eigenvalue 

def power(a,x, max_iteration, tolerable_error):
    lambda_old = 1.0
    condition = True
    step = 1
    while condition:
        # Multiplying a and x
        x = np.matmul(a,x)
        # Finding new Eigen value and Eigen vector
        lambda_new = max(abs(x))
        x = x/lambda_new
        
        # Displaying Eigen value and Eigen Vector
        print('\nIteration %d' %(step))
        print('----------')
        print('Eigen Value = %0.6f' %(lambda_new))
        print('Eigen Vector : ', x)      
        
        # Checking maximum iteration
        step = step + 1
        if step > max_iteration:
            print('Not convergent in given maximum iteration!')
            break
        
        # Calculating error
        error = abs(lambda_new - lambda_old)
        err2 = np.round(error, 7)
        print('Error = '+ str(err2))
        lambda_old = lambda_new
        condition = error > tolerable_error




In [6]:
A = np.array([[4, 2], [2, 4]], float)
X = [1, 0.5]
power(A, X, 100, 1.0e-6)


Iteration 1
----------
Eigen Value = 5.000000
Eigen Vector :  [1.  0.8]
Error = 4.0

Iteration 2
----------
Eigen Value = 5.600000
Eigen Vector :  [1.         0.92857143]
Error = 0.6

Iteration 3
----------
Eigen Value = 5.857143
Eigen Vector :  [1.         0.97560976]
Error = 0.2571429

Iteration 4
----------
Eigen Value = 5.951220
Eigen Vector :  [1.         0.99180328]
Error = 0.0940767

Iteration 5
----------
Eigen Value = 5.983607
Eigen Vector :  [1.         0.99726027]
Error = 0.032387

Iteration 6
----------
Eigen Value = 5.994521
Eigen Vector :  [1.         0.99908592]
Error = 0.010914

Iteration 7
----------
Eigen Value = 5.998172
Eigen Vector :  [1.         0.99969521]
Error = 0.0036513

Iteration 8
----------
Eigen Value = 5.999390
Eigen Vector :  [1.         0.99989839]
Error = 0.0012186

Iteration 9
----------
Eigen Value = 5.999797
Eigen Vector :  [1.         0.99996613]
Error = 0.0004064

Iteration 10
----------
Eigen Value = 5.999932
Eigen Vector :  [1.         0.99998

In [66]:
# using power method to find the dominant eigenvalue

power(a,x, itermax, errmax)


Iteration 1
----------
Eigen Value = 40.000000
Eigen Vector :  [ 0.  0. -1.]
Error = 39.0

Iteration 2
----------
Eigen Value = 50.000000
Eigen Vector :  [-0.4  1.   0.2]
Error = 10.0

Iteration 3
----------
Eigen Value = 60.000000
Eigen Vector :  [ 0.43333333  0.43333333 -1.        ]
Error = 10.0

Iteration 4
----------
Eigen Value = 71.666667
Eigen Vector :  [-0.4         1.         -0.04186047]
Error = 11.6666667

Iteration 5
----------
Eigen Value = 57.581395
Eigen Vector :  [ 0.36752827  0.66155089 -1.        ]
Error = 14.0852713

Iteration 6
----------
Eigen Value = 80.137318
Eigen Vector :  [-0.30460639  1.         -0.19625038]
Error = 22.5559229

Iteration 7
----------
Eigen Value = 54.129624
Eigen Vector :  [ 0.28105099  0.86397155 -1.        ]
Error = 26.0076942

Iteration 8
----------
Eigen Value = 87.369372
Eigen Vector :  [-0.22653035  1.         -0.31564331]
Error = 33.2397477

Iteration 9
----------
Eigen Value = 53.516862
Eigen Vector :  [ 0.19588301  1.         -0.959

With a tolerable error of $10^{-6}$, the solution converges on the 100th iteration with : 
- Eigen Value = 70.943484
- Eigen Vector : [-0.024524	, 1.000000	, -0.623774]

In [67]:
b = a 
pprint(b)

from numpy.linalg import inv

a_inv = inv(b)
pprint(a_inv)

array([[-30.,  10.,  20.],
       [ 10.,  40., -50.],
       [ 20., -50., -10.]])
array([[-0.05576923, -0.01730769, -0.025     ],
       [-0.01730769, -0.00192308, -0.025     ],
       [-0.025     , -0.025     , -0.025     ]])


In [68]:
# Inverse Power Method code to find the smallest eigenvalue

y = np.array([1,1,1], float)

def inv_power(a_inv, x, tolerable_error, max_iteration):
    lambda_old = 1.0
    condition = True
    step = 1
    while condition:
        x = np.dot(a_inv,x)
        lambda_new = min(x)
        x = x/lambda_new
    
        # Displaying Eigen value and Eigen Vector
        print ('\nIteration %d' %(step))
        print ('----------')
        print ('Eigen Value = %0.6f' %(1/lambda_new))
        print ('Eigen Vector : ', x)
    
        # Checking maximum iteration
        step = step + 1
        if step > max_iteration:
            print ('Not convergent in given maximum iteration!')
            break
        
        # Calculating error
        error = abs((1/lambda_new) - (1/lambda_old))
        err2 = np.round(error, 7)
        print ('Error = '+ str(err2))
        lambda_old = lambda_new
        condition = error > tolerable_error




In [69]:
# Using Inverse Power Method to find the smallest eigenvalue

inv_power(a_inv, y, errmax, itermax)


Iteration 1
----------
Eigen Value = -10.196078
Eigen Vector :  [1.         0.45098039 0.76470588]
Error = 11.1960784

Iteration 2
----------
Eigen Value = -12.093023
Eigen Vector :  [1.         0.45098039 0.66985864]
Error = 1.8969448

Iteration 3
----------
Eigen Value = -12.450025
Eigen Vector :  [1.         0.43477264 0.66011245]
Error = 0.3570013

Iteration 4
----------
Eigen Value = -12.531807
Eigen Vector :  [1.         0.43418456 0.65631738]
Error = 0.0817822

Iteration 5
----------
Eigen Value = -12.548327
Eigen Vector :  [1.         0.43355219 0.65580755]
Error = 0.0165203

Iteration 6
----------
Eigen Value = -12.552058
Eigen Vector :  [1.         0.43350587 0.65564414]
Error = 0.0037314

Iteration 7
----------
Eigen Value = -12.552829
Eigen Vector :  [1.         0.43348006 0.65561854]
Error = 0.00077

Iteration 8
----------
Eigen Value = -12.553000
Eigen Vector :  [1.         0.43347732 0.65561136]
Error = 0.0001712

Iteration 9
----------
Eigen Value = -12.553036
Eigen Ve

With a tolerable error of $10^{-6}$, the solution converges on the $12^{th}$ iteration with : 
- Eigen Value = -12.553045
- Eigen Vector :  [1.       ,  0.43347603 , 0.65560971]

In [70]:
# Problem 2

# creating the matrix

A = 4 * np.eye(10)
A = np.array(A, float)

B = np.array(np.zeros((10, 10)), float)
C = np.array(np.zeros((10, 10)), float)

for i in range(9):
    B[i][i + 1] = 2
    B[i + 1][i] = 2
    
for i in range(8):
    C[i][i + 2] = 1
    C[i + 2][i] = 1

M = A + B + C    

N = M

M_inv = inv(N)

pprint(M)
print("\n")
pprint(M_inv)

array([[4., 2., 1., 0., 0., 0., 0., 0., 0., 0.],
       [2., 4., 2., 1., 0., 0., 0., 0., 0., 0.],
       [1., 2., 4., 2., 1., 0., 0., 0., 0., 0.],
       [0., 1., 2., 4., 2., 1., 0., 0., 0., 0.],
       [0., 0., 1., 2., 4., 2., 1., 0., 0., 0.],
       [0., 0., 0., 1., 2., 4., 2., 1., 0., 0.],
       [0., 0., 0., 0., 1., 2., 4., 2., 1., 0.],
       [0., 0., 0., 0., 0., 1., 2., 4., 2., 1.],
       [0., 0., 0., 0., 0., 0., 1., 2., 4., 2.],
       [0., 0., 0., 0., 0., 0., 0., 1., 2., 4.]])


array([[ 0.34601107, -0.17788994, -0.02826441,  0.07606643, -0.02930641,
        -0.01123413,  0.01582546, -0.00416802, -0.00319114,  0.00263758],
       [-0.17788994,  0.43744709, -0.16333442, -0.06733963,  0.0910127 ,
        -0.02344513, -0.01914686,  0.01738847, -0.00231195, -0.00319114],
       [-0.02826441, -0.16333442,  0.43972647, -0.16958645, -0.06479974,
         0.09182677, -0.02500814, -0.01810485,  0.01738847, -0.00416802],
       [ 0.07606643, -0.06733963, -0.16958645,  0.45639857, -0.175

In [78]:
X = np.zeros(10) + 1
X[-1] += 1
power(M, X, itermax, 0.00001)


Iteration 1
----------
Eigen Value = 11.000000
Eigen Vector :  [0.63636364 0.81818182 0.90909091 0.90909091 0.90909091 0.90909091
 0.90909091 1.         1.         1.        ]
Error = 10.0

Iteration 2
----------
Eigen Value = 9.727273
Eigen Vector :  [0.52336449 0.74766355 0.88785047 0.92523364 0.93457944 0.94392523
 0.96261682 1.         0.91588785 0.71962617]
Error = 1.2727273

Iteration 3
----------
Eigen Value = 9.588785
Eigen Vector :  [0.4668616  0.70272904 0.87134503 0.94249513 0.97270955 0.99025341
 1.         0.98245614 0.8411306  0.59551657]
Error = 0.1384877

Iteration 4
----------
Eigen Value = 9.831384
Eigen Vector :  [0.42153267 0.65401011 0.83563002 0.93080202 0.97928026 1.
 0.99266382 0.93556062 0.76494498 0.51333399]
Error = 0.242599

Iteration 5
----------
Eigen Value = 9.810251
Eigen Vector :  [0.39038563 0.61784025 0.80660091 0.91812523 0.97928372 1.
 0.97714136 0.89404382 0.70846639 0.46061886]
Error = 0.0211332

Iteration 6
----------
Eigen Value = 9.725019
Eige

We can see that we found the dominant eigenvalue using power method in 62 iterations
- Eigen Value = 9.555019
- Eigen Vector :  [0.34685395 0.57265926 0.78147769 0.9244928  0.99999089 1.
 0.92451739 0.7815107  0.57269042 0.34687601]

I also noticed that the number of iterations taken for the eigenvalue to dominate are different when we take different initial guess vectors. 

In [72]:
Y = np.zeros(10) + 1
Y[-1] += 1
inv_power(M_inv, Y, errmax, 4000)


Iteration 1
----------
Eigen Value = -11.273862
Eigen Vector :  [-2.13215859 -0.9695301  -0.8061674  -1.51908957 -0.93979442 -0.73604993
 -2.19823789 -0.53450808  1.         -6.00330396]
Error = 12.273862

Iteration 2
----------
Eigen Value = -0.416965
Eigen Vector :  [ 0.28095966 -0.07109468 -0.0926147   0.3119485  -0.05606527 -0.24593286
  0.57670546 -0.06259071 -0.71712215  1.        ]
Error = 10.8568974

Iteration 3
----------
Eigen Value = -1.819542
Eigen Vector :  [-0.28189822  0.18188096  0.25261309 -0.53959384  0.15538793  0.59288899
 -0.87874201  0.06223118  1.         -0.97044327]
Error = 1.4025774

Iteration 4
----------
Eigen Value = -1.364272
Eigen Vector :  [ 0.28542042 -0.19607299 -0.36494973  0.69521542 -0.16853943 -0.78165744
  1.         -0.04036889 -0.9885645   0.8353617 ]
Error = 0.4552696

Iteration 5
----------
Eigen Value = -1.371560
Eigen Vector :  [-0.3180755   0.19338153  0.49406786 -0.85658504  0.16876138  0.95377082
 -1.13232307  0.04071486  1.         -0.7

 -0.13696455 -0.67653445  0.72615612 -0.27591206]
Error = 0.0287672

Iteration 370
----------
Eigen Value = -1.202189
Eigen Vector :  [-0.27053377  0.72740405 -0.69400181 -0.11642542  1.         -1.01182245
  0.13843233  0.68466156 -0.73475027  0.27913436]
Error = 0.028341

Iteration 371
----------
Eigen Value = -1.174258
Eigen Vector :  [ 0.26745788 -0.71901949  0.68588396  0.11523416 -0.98848578  1.
 -0.13666731 -0.67678723  0.72617418 -0.27583423]
Error = 0.0279309

Iteration 372
----------
Eigen Value = -1.201775
Eigen Vector :  [-0.27061228  0.72738584 -0.69374684 -0.11672523  1.         -1.01147681
  0.13808874  0.68467967 -0.73451728  0.27896142]
Error = 0.0275172

Iteration 373
----------
Eigen Value = -1.174657
Eigen Vector :  [ 0.26762441 -0.71924385  0.68586653  0.11556502 -0.98881861  1.
 -0.13637861 -0.67703275  0.72619172 -0.27575863]
Error = 0.0271188

Iteration 374
----------
Eigen Value = -1.201374
Eigen Vector :  [-0.27068847  0.72736816 -0.69349937 -0.11701622  1.   

  0.12682846  0.68527301 -0.72688163  0.27329371]
Error = 0.00036

Iteration 667
----------
Eigen Value = -1.187863
Eigen Vector :  [ 0.27314541 -0.72668184  0.68528854  0.12653383 -0.99985291  1.
 -0.12680763 -0.68517233  0.72677324 -0.27325241]
Error = 0.0003547

Iteration 668
----------
Eigen Value = -1.188213
Eigen Vector :  [-0.27318608  0.72678863 -0.68538776 -0.12655431  1.         -1.00014496
  0.12682414  0.68527324 -0.72687871  0.27329153]
Error = 0.0003495

Iteration 669
----------
Eigen Value = -1.187868
Eigen Vector :  [ 0.27314755 -0.72668473  0.68528832  0.12653808 -0.99985719  1.
 -0.12680392 -0.68517549  0.72677347 -0.27325144]
Error = 0.0003444

Iteration 670
----------
Eigen Value = -1.188208
Eigen Vector :  [-0.27318704  0.72678841 -0.68538466 -0.12655796  1.         -1.00014074
  0.12681995  0.68527346 -0.72687587  0.27328943]
Error = 0.0003394

Iteration 671
----------
Eigen Value = -1.187873
Eigen Vector :  [ 0.27314963 -0.72668752  0.6852881   0.1265422  -0.9998

Eigen Value = -1.188038
Eigen Vector :  [ 0.2732187  -0.72678058  0.68528087  0.12667944 -0.99999939  1.
 -0.12668057 -0.68528039  0.72678096 -0.27321914]
Error = 1.5e-06

Iteration 1040
----------
Eigen Value = -1.188040
Eigen Vector :  [-0.27321887  0.72678103 -0.68528128 -0.12667952  1.         -1.0000006
  0.12668064  0.6852808  -0.7267814   0.27321931]
Error = 1.4e-06

Iteration 1041
----------
Eigen Value = -1.188038
Eigen Vector :  [ 0.27321871 -0.7267806   0.68528087  0.12667946 -0.99999941  1.
 -0.12668056 -0.6852804   0.72678096 -0.27321914]
Error = 1.4e-06

Iteration 1042
----------
Eigen Value = -1.188040
Eigen Vector :  [-0.27321887  0.72678103 -0.68528126 -0.12667954  1.         -1.00000058
  0.12668062  0.6852808  -0.72678139  0.2732193 ]
Error = 1.4e-06

Iteration 1043
----------
Eigen Value = -1.188038
Eigen Vector :  [ 0.27321872 -0.72678061  0.68528086  0.12667947 -0.99999943  1.
 -0.12668054 -0.68528041  0.72678096 -0.27321914]
Error = 1.4e-06

Iteration 1044
------

*Using inverse power method to find smallest eigenvalue in this case was expensive and took 3243 iterations for the solution to* *converge. The smallest eigenvalue was found to be -1.188040*
- *Eigen Value = -1.188040*
- *Eigen Vector :  [-0.27321891, 0.72678102, -0.68528114, -0.12667969,  1. , -1.00000041, 0.12668045,  0.68528081, -0.72678127, 0.27321921]*

*Hence, we can see the usefulness of Power method in obtaining the dominant and smallest eigenvalues.*